dbms/src/Interpreters/ActionsVisitor.cpp

```diff
@@ -587,18 +608,53 @@ void ActionsMatcher::visit(const ASTFunction & node, const ASTPtr & ast, Data &
     }
 }
 
-void ActionsMatcher::visit(const ASTLiteral & literal, const ASTPtr & ast, Data & data)
+void ActionsMatcher::visit(const ASTLiteral & literal, const ASTPtr & /* ast */,
+    Data & data)
 {
-    CachedColumnName column_name;
-    if (data.hasColumn(column_name.get(ast)))
-        return;
-
     DataTypePtr type = applyVisitor(FieldToDataType(), literal.value);
+    const auto value = convertFieldToType(literal.value, *type);
+
+    // FIXME why do we have a second pass with a clean sample block over the same
+    // AST here? Anyway, do not modify the column name if it is set already.
+    if (literal.unique_column_name.empty())
+    {
+        const auto default_name = literal.getColumnName();
+        auto & block = data.getSampleBlock();
+        auto * existing_column = block.findByName(default_name);
+
+        /*
+         * To approximate CSE, bind all identical literals to a single temporary
+         * columns. We try to find the column by its default name, but after that
+         * we have to check that it contains the correct data. This might not be
+         * the case if it is a user-supplied column, or it is from under a join,
+         * etc.
+         * Overall, this is a hack around a generally poor name-based notion of
+         * column identity we currently use.
+         */
+        if (existing_column
+            && existing_column->column
+            && isColumnConst(*existing_column->column)
+            && existing_column->column->size() == 1
+            && existing_column->column->get(0) == value)
```

Comment:
I think it's better to use `checkAndGetColumn<ColumnConst>(existing_column->column)` and its `Field operator[]` instead of adding get() method to IColumn.

Reply:
I can as well use `operator[]` for the `IColumn`, and `get` just calls this operator. Don't remember why I added it, maybe it seemed ugly to dereference. Let's keep it, having to dereference or to write `column->operator[](0)` is somewhat awkward.